packetbeat/protos/cassandra/parser.go

```diff
@@ -0,0 +1,266 @@
+		Message: applayer.Message{
+			Ts: ts,
+		},
+	}
+}
+
+func (p *parser) parserBody() (map[string]interface{}, error) {
+
+	if p.framer.Header.BodyLength > 0 {
+		debugf("bodyLength: %d", p.framer.Header.BodyLength)
+
+		//let's wait for enough buf
+		if !p.buf.Avail(p.framer.Header.BodyLength) {
+			if isDebug {
+				debugf("buf not enough for body, waiting for more, return")
+			}
+			return nil, nil
+		}
+
+		//check if the ops already ignored
+		if p.message.ignored {
+			debugf("message marked to be ignored, let's do this")
+			p.buf.Collect(p.framer.Header.BodyLength)
+			finalCollectedFrameLength := p.buf.BufferConsumed()
+			if finalCollectedFrameLength-p.framer.Header.HeadLength != p.framer.Header.BodyLength {
+				return nil, errors.New("data messed while parse frame body")
+			}
+
+		}
+
+		// start to parse body
+		if !p.message.ignored {
+			data, err := p.framer.ReadFrame()
+			if err != nil {
+				// if the frame parsed failed, should ignore the whole message
+				p.framer = nil
+				return nil, err
+			}
+
+			// dealing with un-parsed content
+			frameParsedLength := p.buf.BufferConsumed()
+
+			// collect leftover
+			unParsedSize := p.framer.Header.BodyLength + p.framer.Header.HeadLength - frameParsedLength
+			if unParsedSize > 0 {
+
+				// double check the buf size
+				if p.buf.Avail(unParsedSize) {
+					p.buf.Collect(unParsedSize)
+				} else {
+					logp.Err("should be enough bytes for cleanup,but not enough")
+					return nil, errors.New("should be enough bytes,but not enough")
+				}
+			}
+			return data, nil
+		}
+
+		finalCollectedFrameLength := p.buf.BufferConsumed()
+		if finalCollectedFrameLength-p.framer.Header.HeadLength != p.framer.Header.BodyLength {
+			logp.Err("body_length:%d, head_length:%d, all_consumed:%d", p.framer.Header.BodyLength, p.framer.Header.HeadLength, finalCollectedFrameLength)
+			return nil, errors.New("data messed while parse frame body")
+		}
+
+	}
+
+	return map[string]interface{}{}, nil
+}
+
+func (p *parser) parse() (*message, error) {
+
+	// if p.frame is nil then create a new framer, or continue to process the last message
+	if p.framer == nil {
+		if isDebug {
+			debugf("start new framer")
+		}
+		p.framer = NewFramer(&p.buf, p.config.compressor)
+	}
+
+	// check if the frame header were parsed or not
+	if p.framer.Header == nil {
+		if isDebug {
+			debugf("start to parse header")
+		}
+		if !p.buf.Avail(9) {
+			debugf("not enough head bytes, ignore")
+			return nil, nil
+		}
+
+		_, err := p.framer.ReadHeader()
+		if err != nil {
+			logp.Err("%v", err)
+			p.framer = nil
+			return nil, err
+		}
+	}
+
+	//check if the ops need to be ignored
+	if p.FrameOpsIgnored() {
+		// as we already ignore the content, we now mark the result is ignored
+		p.message.ignored = true
+		if isDebug {
+			debugf("Ops: %s was marked to be ignored, ignoring, request:%v", p.framer.Header.Op.String(), p.framer.Header.Version.IsRequest())
+		}
+	}
+
+	msg := p.message
+
+	data, err := p.parserBody()
```

Comment:
consider parserBody to return an additional bool. Creating a map will allocate an empty map, which might not really be used. With `message.data` already being allocated, parseBody should write into `message.data` and return bool if finished.